lib/msprime.c

```diff
@@ -7253,52 +7224,50 @@ msp_set_simulation_model_wf_ped(msp_t *self)
         goto out;
     }
 
-    parents = malloc(num_individuals * self->ploidy * sizeof(*parents));
-    time = malloc(num_individuals * sizeof(*time));
-    is_sample = malloc(num_individuals * sizeof(*is_sample));
-    if (parents == NULL || time == NULL || is_sample == NULL) {
-        ret = MSP_ERR_NO_MEMORY;
-        goto out;
-    }
-
     for (j = 0; j < num_individuals; j++) {
         ret = tsk_individual_table_get_row(
-            &self->tables->individuals, (tsk_id_t) j, &ind);
+            &self->tables->individuals, (tsk_id_t) j, &ind_obj);
         tsk_bug_assert(ret == 0);
-        ind_parents = ind.parents;
-        tsk_bug_assert(ind.parents_length == self->ploidy);
+        ind = &self->pedigree->inds[j];
+        ind->id = (tsk_id_t) j;
+        ind_parents = ind_obj.parents;
+        tsk_bug_assert(ind_obj.parents_length == self->ploidy);
         for (k = 0; k < self->ploidy; k++) {
             if (ind_parents[k] < TSK_NULL
                 || ind_parents[k] >= (tsk_id_t) num_individuals) {
                 ret = TSK_ERR_INDIVIDUAL_OUT_OF_BOUNDS;
                 goto out;
             }
-            parents[j * self->ploidy + k] = ind_parents[k];
+            if (ind_parents[k] != TSK_NULL) {
+                ind->parents[k] = &self->pedigree->inds[ind_parents[k]];
+            }
+        }
+        sample_flag = false;
+
+        ind_obj.nodes_length = self->ploidy;
+        for (k = 0; k < ind_obj.nodes_length; k++) {
+            // ret = tsk_treeseq_get_node(&ts, ind_obj.nodes[k], &node_obj);
```

Comment:
Why can't you use ``ind_obj.nodes[k]`` here?

Reply:
I think because [tsk_individual_table_get_row](https://github.com/tskit-dev/tskit/blob/main/c/tskit/tables.c#L1133) returns an tsk_individual_t with nodes that are null

Reply:
You want to call ``tsk_treeseq_get_node`` though - that's the point of creating the ``tsk_treeseq_t`` object at the top of the function.

Reply:
`tsk_treeseq_get_node` is still being called to get the node because we need the node time and sample flag

Reply:
Right, sorry - you want to call ``tsk_treeseq_get_individual`` at the top of the loop above. That'll populate the ``nodes`` field.

Reply:
thanks, I'll try that!